hail/src/main/scala/is/hail/utils/StackSafe.scala

```diff
@@ -0,0 +1,176 @@
+package is.hail.utils
+
+import scala.annotation.tailrec
+import scala.collection.generic.{CanBuild, CanBuildFrom}
+
+object StackSafe {
+
+  def done[A](result: A): StackFrame[A] = Done(result)
+
+  def call[A](body: => StackFrame[A]): StackFrame[A] =
+    Thunk(() => body)
+
+  object StackFrame {
```

Comment:
I generally don't like nesting objects, since it generates JVM bytecode that's a bit harder for the JVM to trace to a specific line. Can we move this outside?